src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,765 @@
+            RandomizedResendActiveStatusAfter = RandomizeTimespan(ResendActiveStatusAfter);
+        }
+   
+        // as a backup measure, current local active status is sent occasionally
+        public static TimeSpan ResendActiveStatusAfter = new TimeSpan(hours: 0, minutes: 10, seconds: 0);
+
+        // time after which this gateway removes other gateways in this same cluster that are known to be gone 
+        public static TimeSpan CleanupSilentGoneGatewaysAfter = new TimeSpan(hours: 0, minutes: 0, seconds: 30);
+
+        // to avoid convoying, each silo randomizes these period intervals
+        private TimeSpan RandomizedBackgroundGossipInterval;
+        private TimeSpan RandomizedResendActiveStatusAfter;
+
+        // randomize a timespan by up to 10%
+        private TimeSpan RandomizeTimespan(TimeSpan value)
+        {
+            return TimeSpan.FromMilliseconds(value.TotalMilliseconds * (0.9 + (random.NextDouble() * 0.1)));
+        }
+
+        public bool IsFunctionalClusterGateway(SiloAddress siloAddress)
+        {
+            GatewayEntry g;
+            return localData.Current.Gateways.TryGetValue(siloAddress, out g) 
+                && g.Status == GatewayStatus.Active;
+        }
+
+        public IEnumerable<string> GetActiveClusters()
+        {
+            var clusters = localData.Current.Gateways.Values
+                 .Where(g => g.Status == GatewayStatus.Active)
+                 .Select(g => g.ClusterId)
+                 .Distinct();
+
+            return clusters;
+        }
+
+        public IEnumerable<GatewayEntry> GetGateways()
+        {
+            return localData.Current.Gateways.Values;
+        }
+
+        public SiloAddress GetRandomClusterGateway(string cluster)
+        {
+            var activegateways = new List<SiloAddress>();
+
+            foreach(var gw in localData.Current.Gateways)
+            {
+                var cur = gw.Value;
+                if (cur.ClusterId != cluster)
+                    continue;
+                if (cur.Status != GatewayStatus.Active)
+                    continue;
+                activegateways.Add(cur.SiloAddress);
+            }
+
+            if (activegateways.Count == 0)
+                return null;
+
+            return activegateways[random.Next(activegateways.Count)];
+        }
+
+        public MultiClusterConfiguration GetMultiClusterConfiguration()
+        {
+            return localData.Current.Configuration;
+        }
+
+        public async Task InjectMultiClusterConfiguration(MultiClusterConfiguration config)
+        {
+            this.injectedConfig = config;
+            PushChanges();
+
+            // wait for the gossip channel tasks and reproduce any exceptions
+            foreach (var ct in this.channelTasks.Values)
+            {
+                await ct.Task;
+                if (ct.LastException != null)
+                    throw ct.LastException;
+            }
```

Comment:
I do not follow all this code yet, but wouldn't it make sense to do `await Task.WhenAll(this.channelTasks.Values.Select(ct => ct.Task)` first? Otherwise the exception might interrupt awaiting all of them. Maybe this is fine, but just checking with you.